File: fixtures/scale/objc/large/Families/Substitution/SubstitutionCipher.m
#import "SubstitutionCipher.h"

@implementation SubstitutionCipher

- (instancetype)init {
    if ((self = [super init])) {
        _shift = 6; _step = 2;
    }
    return self;
}

- (NSString *)name {
    return @"substitution";
}

- (NSString *)encode:(NSString *)plaintext {
    NSMutableString *out = [NSMutableString string];
    for (NSUInteger i = 0; i < plaintext.length; i++) {
        unichar ch = [plaintext characterAtIndex:i];
        if (ch < 'A' || ch > 'Z') { [out appendFormat:@"%C", ch]; continue; }
        NSInteger shifted = ((ch - 'A') + self.shift + (NSInteger)i * self.step) % 26;
        [out appendFormat:@"%c", (char)('A' + shifted)];
    }
    return out;
}

- (NSString *)decode:(NSString *)ciphertext {
    NSMutableString *out = [NSMutableString string];
    for (NSUInteger i = 0; i < ciphertext.length; i++) {
        unichar ch = [ciphertext characterAtIndex:i];
        if (ch < 'A' || ch > 'Z') { [out appendFormat:@"%C", ch]; continue; }
        NSInteger shifted = ((ch - 'A') - self.shift - (NSInteger)i * self.step + 2600) % 26;
        [out appendFormat:@"%c", (char)('A' + shifted)];
    }
    return out;
}

@end

File: fixtures/scale/php/large/src/Families/Substitution/SubstitutionCipher.php
<?php

namespace CipherLab\Families\Substitution;

use CipherLab\Abstractions\Cipher;

class SubstitutionCipher implements Cipher
{
    private int $shift; private int $step;

    public function __construct()
    {
        $this->shift = 6; $this->step = 2;
    }

    public function name(): string
    {
        return 'substitution';
    }

    public function encode(string $plaintext): string
    {
        $chars = str_split($plaintext);
        foreach ($chars as $i => $ch) {
            if ($ch < 'A' || $ch > 'Z') { continue; }
            $chars[$i] = chr(65 + (ord($ch) - 65 + $this->shift + $i * $this->step) % 26);
        }
        return implode('', $chars);
    }

    public function decode(string $ciphertext): string
    {
        $chars = str_split($ciphertext);
        foreach ($chars as $i => $ch) {
            if ($ch < 'A' || $ch > 'Z') { continue; }
            $chars[$i] = chr(65 + ((ord($ch) - 65 - $this->shift - $i * $this->step) % 26 + 2600) % 26);
        }
        return implode('', $chars);
    }
}
